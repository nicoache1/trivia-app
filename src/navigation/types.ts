import { ParamListBase } from '@react-navigation/native'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React from 'react'

import { Routes } from './routes'

export interface ParamList extends ParamListBase {
  [Routes.Welcome]: undefined
  [Routes.ConfigureGame]: undefined
  [Routes.Game]: undefined
  [Routes.Results]: undefined
  [Routes.AnimatedSplash]: undefined
}

export type SceneProps<T extends keyof ParamList> = React.FC<
  NativeStackScreenProps<ParamList, T>
>

export type NavigationProp<T extends keyof ParamList> = NativeStackScreenProps<
  ParamList,
  T
>['navigation']
