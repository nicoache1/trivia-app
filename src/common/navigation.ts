import {
  NavigationContainerRef,
  ParamListBase,
  StackActions,
} from '@react-navigation/native'
import * as React from 'react'

import { ParamList } from '../navigation/types'

export const navigationRef: React.RefObject<NavigationContainerRef<ParamList>> =
  React.createRef()

export function navigate(name: string, params?: ParamListBase[string]) {
  navigationRef.current?.navigate(name, params)
}

export function push(name: string, params?: ParamListBase[string]) {
  navigationRef.current?.dispatch(StackActions.push(name, params))
}

export function popToTop() {
  navigationRef.current?.dispatch(StackActions.popToTop())
}

export function pop() {
  navigationRef.current?.dispatch(StackActions.pop())
}
