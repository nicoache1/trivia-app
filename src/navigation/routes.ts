export enum Routes {
  Game = 'Game',
  ConfigureGame = 'ConfigureGame',
  AnimatedSplash = 'AnimatedSplash',
  Results = 'Results',
  Welcome = 'Welcome',
}
