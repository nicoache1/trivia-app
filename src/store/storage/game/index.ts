import { createSlice } from '@reduxjs/toolkit'

import { restartGame, setGame, setGameResult } from './actions'
import { initialState } from './types'

const gameSlice = createSlice({
  initialState,
  name: 'game',
  reducers: {
    restartGameAction: restartGame,
    setGameAction: setGame,
    setGameResultAction: setGameResult,
  },
})

export const {
  actions: { restartGameAction, setGameAction, setGameResultAction },
  reducer: gameReducer,
} = gameSlice
