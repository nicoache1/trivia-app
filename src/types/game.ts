import { QuestionResult } from './questions'

export enum Difficulties {
  Easy = 'easy',
  Medium = 'medium',
  Hard = 'hard',
}

export enum QuestionAmounts {
  Five = 5,
  Ten = 10,
  TwentyFive = 25,
}

export const getDifficulties = () => Object.values(Difficulties)
export const getQuestionAmounts = () =>
  Object.values(QuestionAmounts).filter(Number) as QuestionAmounts[]

export interface Game {
  player: string
  difficulty: Difficulties
  questionAmount: QuestionAmounts
  results: QuestionResult[]
}
