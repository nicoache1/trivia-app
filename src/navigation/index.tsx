import { createNativeStackNavigator } from '@react-navigation/native-stack'
import React from 'react'

import { AnimatedSplash } from '../scenes/animatedSplash'
import { ConfigureGame } from '../scenes/configureGame'
import { Game } from '../scenes/game'
import { Results } from '../scenes/results'
import { Welcome } from '../scenes/welcome'
import { Routes } from './routes'
import { ParamList } from './types'

const Stack = createNativeStackNavigator<ParamList>()

export const AppNavigator: React.FC<{}> = ({}) => {
  return (
    <Stack.Navigator
      screenOptions={{
        contentStyle: {
          backgroundColor: 'white',
        },
      }}>
      <Stack.Screen name={Routes.AnimatedSplash} component={AnimatedSplash} />
      <Stack.Screen
        name={Routes.Welcome}
        component={Welcome}
        options={{ presentation: 'modal' }}
      />
      <Stack.Screen name={Routes.ConfigureGame} component={ConfigureGame} />
      <Stack.Screen name={Routes.Game} component={Game} />
      <Stack.Screen name={Routes.Results} component={Results} />
    </Stack.Navigator>
  )
}
