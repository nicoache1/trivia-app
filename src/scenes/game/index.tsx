import CheckImage from 'assets/check.png'
import CrossImage from 'assets/cross.png'
import React, { useEffect, useLayoutEffect, useRef, useState } from 'react'
import { View } from 'react-native'
import { useSelector } from 'react-redux'
import { Background } from 'src/components/Background'
import { SceneContainer } from 'src/components/SceneContainer'
import { StyledContainer } from 'src/components/StyledContainer'
import { StyledText } from 'src/components/StyledText'
import { Routes } from 'src/navigation/routes'
import { SceneProps } from 'src/navigation/types'
import { useAppDispatch } from 'src/store'
import { useLazyQuestionsQuery } from 'src/store/APIs/questions'
import { setGameResultAction } from 'src/store/storage/game'
import { selectGameConfiguration } from 'src/store/storage/game/selectors'
import { useTheme } from 'src/styles/Theme'
import { AppTypography, ColorPalette } from 'src/styles/types'
import { Question } from 'src/types/questions'
import { v4 as uuidv4 } from 'uuid'

import { Card } from './Card'
import { FinishGame } from './FinishGame'
import { GameButton } from './GameButton'
import { styles } from './styles'

export const Game: SceneProps<Routes.Game> = ({ navigation }) => {
  const game = useSelector(selectGameConfiguration)
  const {
    Theme: { colors },
  } = useTheme()
  const dispatch = useAppDispatch()
  const [triggerQuery, { isLoading, isFetching }] = useLazyQuestionsQuery()

  const ref = useRef<{
    triggerAnimation: (option: boolean) => void
  }>(null)

  const [cards, setCards] = useState<Question[]>([])

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    })
  }, [navigation])

  useEffect(() => {
    async function getQuestions() {
      const data = await triggerQuery({
        amount: game.questionAmount,
        difficulty: game.difficulty,
      }).unwrap()
      setCards(data)
    }

    getQuestions()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const onFinish = (option: boolean) => {
    const question = cards[cards.length - 1]

    dispatch(
      setGameResultAction({
        results: [
          {
            correct: question.correctAnswer === option,
            id: uuidv4(),
            question,
          },
        ],
      }),
    )

    setCards(prevState => {
      return prevState.slice(0, prevState.length - 1)
    })
  }

  const onPressButton = (option: boolean) => () => {
    ref.current?.triggerAnimation(option)
  }

  const finishGame = cards.length === 0 && !isLoading && !isFetching

  const finishCallback = () => navigation.navigate(Routes.Results)

  return (
    <>
      {finishGame && <FinishGame callback={finishCallback} />}
      <SceneContainer
        barStyle={'light-content'}
        style={[styles.container, { backgroundColor: colors.SECONDARY }]}>
        <Background />
        <StyledContainer style={styles.playerContainer}>
          <StyledText color={ColorPalette.WHITE} typography={AppTypography.H1}>
            {game.player}
          </StyledText>
        </StyledContainer>
        <View style={styles.cardContainer}>
          {cards.map((item, index) => (
            <Card
              ref={index === cards.length - 1 ? ref : undefined}
              key={item.id}
              item={item}
              index={index}
              onFinish={onFinish}
            />
          ))}
        </View>
        <StyledContainer style={styles.answerOptionContainer}>
          <GameButton imageSource={CheckImage} onPress={onPressButton(true)} />
          <GameButton imageSource={CrossImage} onPress={onPressButton(false)} />
        </StyledContainer>
      </SceneContainer>
    </>
  )
}
