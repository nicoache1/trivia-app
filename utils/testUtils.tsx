import type { PreloadedState } from '@reduxjs/toolkit'
import type { RenderOptions } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { AppStore, RootState, setupStore } from 'src/store'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProvider(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
  }: ExtendedRenderOptions = {},
) {
  return <Provider store={store}>{ui}</Provider>
}
