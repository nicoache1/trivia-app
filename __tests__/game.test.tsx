import { render } from '@testing-library/react-native'
import React from 'react'
import { Routes } from 'src/navigation/routes'
import { Game } from 'src/scenes/game'

import {
  createNavigationPropMock,
  createRoutePropMock,
} from '../__mocks__/navigationMocks'
import { renderWithProvider } from '../utils/testUtils'

test('Expect Game to be rendered correctly with default', async () => {
  const mockRoute = createRoutePropMock<Routes.Game>(Routes.Game)

  const mockNavigation = createNavigationPropMock<Routes.Game>()

  const enhancedComponent = renderWithProvider(
    <Game route={mockRoute} navigation={mockNavigation} />,
  )

  const renderedSnapshot = render(enhancedComponent).toJSON()

  expect(renderedSnapshot).toMatchSnapshot()
})
