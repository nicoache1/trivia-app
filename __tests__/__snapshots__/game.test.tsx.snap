// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect Game to be rendered correctly with default 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
    ]
  }
  style={
    Array [
      Object {
        "flexGrow": 1,
      },
      Object {
        "backgroundColor": "#FFFFFF",
      },
      Array [
        Object {
          "flex": 1,
          "paddingVertical": 60,
        },
        Object {
          "backgroundColor": "rgb(108,92,231)",
        },
      ],
    ]
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 70,
      }
    }
  >
    <RNSVGSvgView
      align="xMinYMin"
      bbHeight="812"
      bbWidth={750}
      fill="none"
      focusable={false}
      height="812"
      meetOrSlice={1}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 812,
            "width": 750,
          },
        ]
      }
      vbHeight={812}
      vbWidth={375}
      width={750}
    >
      <RNSVGGroup
        fill={null}
        propList={
          Array [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M-757.5 612.5L-1210 243.5V948.5H459V264.5L112 673L-335.5 -76L-757.5 612.5Z"
          fill={4279510628}
          fillOpacity={0.3}
          propList={
            Array [
              "fill",
              "fillOpacity",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 100,
        "justifyContent": "center",
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      accessible={true}
      color="WHITE"
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 32,
          "fontStyle": "normal",
          "fontWeight": "700",
        }
      }
      typography="H1"
    >
      Player
    </Text>
  </View>
  <View
    style={
      Object {
        "marginHorizontal": 10,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "bottom": 66.7,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "left": 0,
        "marginHorizontal": 15,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "opacity": 1,
          },
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 40,
            "elevation": 11,
            "height": 110,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 5,
              "width": 0,
            },
            "shadowOpacity": 0.36,
            "shadowRadius": 6.68,
            "width": 110,
          },
        ]
      }
      testID="gameButton"
    >
      <Image
        source={1}
        style={
          Object {
            "height": 80,
            "width": 80,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "opacity": 1,
          },
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 40,
            "elevation": 11,
            "height": 110,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 5,
              "width": 0,
            },
            "shadowOpacity": 0.36,
            "shadowRadius": 6.68,
            "width": 110,
          },
        ]
      }
      testID="gameButton"
    >
      <Image
        source={1}
        style={
          Object {
            "height": 80,
            "width": 80,
          }
        }
      />
    </View>
  </View>
</RNCSafeAreaView>
`;
